Add web.dev/monitor-total-page-memory-usage (#2525)
GoogleChrome/web.dev: src/site/content/en/blog/performance-measure-memory/index.md

> ---

Reviewer: @ulan please rename this directory to `monitor-total-memory-usage` or `performance-measurememory`

The first option is more aligned with our policy that the page URL should match the page title, the second option is more aligned with the actual name of the API (having a dash between `measure` and `memory` doesn't make sense to me)

Author: Renamed to `monitor-total-page-memory-usage`

title: Monitor your web page's total memory usage with `performance.measureMemory()`
subhead: >
  Learn how to measure memory usage of your web page in production to detect regressions.
description: >
  Learn how to measure memory usage of your web page in production to detect regressions.
date: 2020-04-07
authors:
  - ulan

> hero: hero.jpg

Reviewer: @ulan please optimize this with [squoosh](squoosh.app) so that we keep the size of the repository down

Author: @kaycebasques thanks for the tip, reduces the size to 116K

---

> title: Monitor your web page's total memory usage with performance.measureMemory()

Reviewer: ```suggestion
title: Monitor your web page's total memory usage with `performance.measureMemory()`
```

subhead: >
  Learn how to measure memory usage of your web page in production to detect regressions.
description: >
  Learn how to measure memory usage of your web page in production to detect regressions.
date: 2020-04-07
authors:
  - ulan
hero: hero.jpg
alt: >
  Green RAM stick. Photo by Harrison Broadbent on Unsplash.
origin_trial:
    url: https://developers.chrome.com/origintrials/#/view_trial/1281274093986906113
tags:
  - post
  - memory
  - javascript

>   - capabilities

Reviewer: ```suggestion
```

This API doesn't seem to be part of the Capabilities project: https://docs.google.com/spreadsheets/d/1de0ZYDOcafNXXwMcg4EZhT0346QM-QFvZfoD8ZffHeA/edit#gid=557099940

garbage collection to detect when an object is no longer needed and to free
the underlying memory chunk. The detection is not perfect though, and it
[was proven][halting-problem] that perfect detection
is an impossible task. Therefore browsers approximate the notion of "an object
is needed" with the notion of "an object is reachable". If the web page cannot
reach an object via its variables and the fields of other reachable objects,
then the browser can safely reclaim the object. The difference between these
two notions leads to memory leaks as illustrated by the following example.

```javascript
const object = { a: new Array(1000), b: new Array(2000) };
setInterval(() => console.log(object.a), 1000);
```

Here the larger array `b` is no longer needed, but the browser does not
reclaim it because it is still reachable via `object.b` in the callback. Thus
the memory of the larger array is leaked.

Memory leaks are [prevalent on the Web][memory-leaks].
It is easy to introduce one by forgetting to unregister an event listener, by
accidentally capturing objects from an iframe, by not closing a worker, by
accumulating objects in arrays, and so on. If a web page has memory leaks,

> then its memory usage grows over time and the web page appears as slow and

Reviewer: ```suggestion
then its memory usage grows over time and the web page appears slow and
```

bloated to the users.

The first step in solving this problem is measuring it. The new

> [performance.measureMemory API][explainer] allows developers to

Reviewer: ```suggestion
[`performance.measureMemory()` API][explainer] allows developers to
```

measure memory usage of their web pages in production and thus detect memory
leaks that slip through local testing.

> ## How is performance.measureMemory() different from the legacy performance.memory API? {: #legacy-api }

Reviewer: ```suggestion
## How is `performance.measureMemory()` different from the legacy `performance.memory` API? {: #legacy-api }
```

different ways of representing objects in memory and different ways of
estimating the memory usage. Browsers may exclude some memory regions from
accounting if proper accounting is too expensive or infeasible. Thus, results
cannot be compared across browsers. It is only meaningful to compare the
results for the same browser.

## Current status {: #status }

<div class="w-table-wrapper">
<table>
    <thead>
      <tr>
        <th>Step</th>
        <th>Status</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>1. Create explainer</td>
        <td><a href="https://github.com/WICG/performance-measure-memory">Complete</a></td>
      </tr>
      <tr>
        <td>2. Create initial draft of specification</td>
        <td>In Progress</td>
      </tr>
      <tr>
        <td>3. Gather feedback and iterate design</td>
        <td><a href="#feedback">In Progress</a></td>
      </tr>
      <tr>
        <td>4. Origin trial </td>
        <td><a href="https://developers.chrome.com/origintrials/#/view_trial/1281274093986906113">In Progress</a></td>
      </tr>
      <tr>
        <td>5. Launch</td>
        <td>Not Started</td>
      </tr>
    </tbody>
</table>
</div>

> ## Using performance.measureMemory() {: use }

Reviewer: ```suggestion
## Using `performance.measureMemory()` {: use }
```

### Enabling support during the origin trial phase

The `performance.measureMemory()` API is available as an origin trial in
Chrome 82. The origin trial is expected to end in Chrome 84.

{% include 'content/origin-trials.njk' %}

### Register for the origin trial {: #register-for-ot }

{% include 'content/origin-trial-register.njk' %}

### Enabling via chrome://flags

To experiment with `performance.measureMemory()` without an origin trial
token, enable the `#experimental-web-platform-features` flag in `chrome://flags`.

### Feature detection

> The `performance.measureMemory` function may fail with a

Reviewer: ```suggestion
The `performance.measureMemory()` function may fail with a
```

      throw error;
    }
  }
  console.log(result);
}
```

### Local testing

Chrome performs the memory measurement during garbage collection. This means
that the API does not resolve the result promise immediately and instead waits
for the next garbage collection. The API forces a garbage collection after
some timeout, which is currently set to 20 seconds. Starting Chrome with the
`--enable-blink-features='ForceEagerMeasureMemory'` command-line flag reduces
the timeout to zero and is useful for local debugging and testing.

## Example

The recommended usage of the API is to define a global memory monitor that
samples memory usage of the whole web page and sends the results to a server
for aggregation and analysis. The simplest way is to sample periodically, for
example every `M` minutes. That however introduces bias to data because the
memory peaks may occur between the samples. The following example shows how to
do unbiased memory measurements using a [Poisson process][poisson], which
guarantees that samples are equally likely to occur at any point in time
([demo][demo], [source][demo-source]).

First, define a function that schedules the next memory measurement using

> setTimeout with a randomized interval. The function should be called after

Reviewer: ```suggestion
`setTimeout()` with a randomized interval. The function should be called after
```

page load on the main window.

```javascript
function scheduleMeasurement() {
  if (!performance.measureMemory) {
    console.log("performance.measureMemory() is not available.");
    return;
  }
  const interval = measurementInterval();
  console.log("Scheduling memory measurement in " +
      Math.round(interval / 1000) + " seconds.");
  setTimeout(performMeasurement, interval);
}

// Start measurements after page load on the main window.
window.onload = function () {
  scheduleMeasurement();
}
```

> The `measurementInterval` function computes a random interval in milliseconds

Reviewer: ```suggestion
The `measurementInterval()` function computes a random interval in milliseconds
```

> such that on average there is one measurement every five minutes. See [this
> article][math] if you are interested in the math behind the function.

Reviewer: ```suggestion
such that on average there is one measurement every five minutes. See [Exponential
distribution][math] if you are interested in the math behind the function.
```

return multiple URLs in `attribution` indicating they could not distinguish
which of these URLs owns the memory.

## Feedback {: #feedback }

The [Web Performance Community Group][webperfs] and the Chrome team would love
to hear about your thoughts and experiences with
`performance.measureMemory()`.

### Tell us about the API design

Is there something about the API that doesn't work as expected? Or are there
missing properties that you need to implement your idea? File a spec issue on
the [performance.measureMemory GitHub repo][issues] or add your thoughts to an
existing issue.

### Report a problem with the implementation

Did you find a bug with Chrome's implementation? Or is the implementation
different from the spec? File a bug at [new.crbug.com][new-bug]. Be sure to
include as much detail as you can, provide simple instructions for reproducing
the bug, and have **Components** set to `Blink>PerformanceAPIs`.
[Glitch][glitch] works great for sharing quick and easy repros.

### Show support

Are you planning to use `performance.measureMemory()`? Your public support
helps the Chrome team prioritize features and shows other browser vendors how

> critical it is to support them. Send a tweet to @ChromiumDev and let us know

Reviewer: ```suggestion
critical it is to support them. Send a tweet to [@ChromiumDev](https://twitter.com/chromiumdev) and let us know
```
